### ISSUE TITLE
s3_bucket versioning parameter is required

### ISSUE THREAD

**Issue by ryanwetzelberger** [ansible/ansible-modules-extras#2736]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

s3_bucket
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

NONE
##### OS / ENVIRONMENT

Mac OS X 10.11.6
##### SUMMARY

When running a play to perform a change other than changing the versioning of the bucket, Ansible will assume that versioning needs to be turned off (default is no).  In all documentation, this parameter is marked as not required.  But if I perform an iteration over many s3 buckets to add something as simple as a tag, some of which need to have versioning enabled (for Cross Region Replication), it will suspend versioning on the buckets that have versioning enabled.  
##### STEPS TO REPRODUCE

```

---

- hosts: localhost
  tasks:
  - name: Tag S3 buckets
    s3_bucket: {name: "{{ item }}", tags:{Application: "S3"}}
    with_items:
      - policy-test-rwetzelberger
```

##### EXPECTED RESULTS

I expect to see that the s3 bucket named "policy-test-rwetzelberger" has a tag verified to have or added called "S3".  This bucket had versioning enabled.  Versioning that is enabled should not be suspended.
##### ACTUAL RESULTS

```
ansible-playbook -vvvv s3-tag.yml
No config file found; using defaults
Loaded callback default of type stdout, v2.0

PLAYBOOK: s3-tag.yml ***********************************************************
1 plays in s3-tag.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: <username>
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1471356881.14-189485368299932 `" && echo ansible-tmp-1471356881.14-189485368299932="` echo $HOME/.ansible/tmp/ansible-tmp-1471356881.14-189485368299932 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/m2/5vh7dqn90gz834mv6swzzdzmh8qf0x/T/tmpiE06ET TO /Users/<username>/.ansible/tmp/ansible-tmp-1471356881.14-189485368299932/setup
<127.0.0.1> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/<username>/.ansible/tmp/ansible-tmp-1471356881.14-189485368299932/setup; rm -rf "/Users/<username>/.ansible/tmp/ansible-tmp-1471356881.14-189485368299932/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [Tag S3 buckets] **********************************************************
task path: /Users/<username>/fork-ansible-playbooks/Gio/tasks/s3-tag.yml:5
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: <username>
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1471356885.59-45084579463130 `" && echo ansible-tmp-1471356885.59-45084579463130="` echo $HOME/.ansible/tmp/ansible-tmp-1471356885.59-45084579463130 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/m2/5vh7dqn90gz834mv6swzzdzmh8qf0x/T/tmpVvlVdL TO /Users/<username>/.ansible/tmp/ansible-tmp-1471356885.59-45084579463130/s3_bucket
<127.0.0.1> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/<username>/.ansible/tmp/ansible-tmp-1471356885.59-45084579463130/s3_bucket; rm -rf "/Users/<username>/.ansible/tmp/ansible-tmp-1471356885.59-45084579463130/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => (item=policy-test-rwetzelberger) => {"changed": true, "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "force": false, "name": "policy-test-rwetzelberger", "policy": null, "profile": null, "region": null, "requester_pays": false, "s3_url": null, "security_token": null, "state": "present", "tags": {"Application": "S3"}, "validate_certs": true, "versioning": false}, "module_name": "s3_bucket"}, "item": "policy-test-rwetzelberger", "name": "policy-test-rwetzelberger", "policy": null, "requester_pays": false, "tags": {"Application": "S3"}, "versioning": {"MfaDelete": "Disabled", "Versioning": "Suspended"}}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

**Comment by ansibot** [comment]
@wimnat ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ryanwetzelberger** [comment]
This also will wipe out any defined policy on an existing bucket.  This should not be modifying resources on an existing bucket if the parameter is undefined.

**Comment by ansibot** [comment]
@wimnat, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by wimnat** [comment]
@ryanwetzelberger try this PR https://github.com/ansible/ansible-modules-extras/pull/2938

**Comment by ryanwetzelberger** [comment]
@wimnat The code looks good, but still need to test.  Trying to get that done today

**Comment by ryanwetzelberger** [comment]
Hello @wimnat.  Very happy with the changes.  There is one more I would recommend though.  I ran the same Ansible script above against a test bucket.  From the script, it is only checking one tag.  The bucket in question has a second tag.  When I run the script, it changes the bucket and removes the second tag that is not called out in the script.  I would not expect this to be the desired behavior of the module.  

Thoughts?

**Comment by ryanwetzelberger** [comment]
adding to that, I could create a new issue against s3_bucket to track that issue if you would prefer to track it separately.

**Comment by ryanwetzelberger** [comment]
https://github.com/wimnat/ansible-modules-extras/blob/feature/issue-2736/cloud/amazon/s3_bucket.py#L242

**Comment by wimnat** [comment]
@ryanwetzelberger for me, it's desired behavior as Ansible is bringing the config in to line with what is defined.  However, i see many occasions where people dont want this.  It's resolved in other modules by adding a 'purge_tags' parameter.

I can do this but will do it separately.  if you could create a seperate issue that'd be great.

**Comment by ryanwetzelberger** [comment]
@wimnat will do!  Without this change, the PR linked to this issue is good!

**Comment by ryanwetzelberger** [comment]
accidentally closed it before PR has been completed. Reopening

**Comment by ryansb** [comment]
Fix for this issue has been merged, thanks for reporting this issue
